vortex-serde/src/layouts/read/selection.rs

```diff
@@ -0,0 +1,164 @@
+        write!(f, "RowSelector [{}..{}]", self.begin, self.end)
+    }
+}
+
+impl RowSelector {
+    pub fn new(values: Bitmap, begin: usize, end: usize) -> Self {
+        Self { values, begin, end }
+    }
+
+    pub fn from_array(array: &Array, begin: usize, end: usize) -> VortexResult<Self> {
+        array.with_dyn(|a| {
+            a.as_bool_array()
+                .ok_or_else(|| vortex_err!("Must be a bool array"))
+                .map(|b| {
+                    let mut bitmap = Bitmap::new();
+                    for (sb, se) in b.maybe_null_slices_iter() {
+                        bitmap.add_range((sb + begin) as u32..(se + begin) as u32);
+                    }
+                    RowSelector::new(bitmap, 0, end)
+                })
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
+
+    pub fn begin(&self) -> usize {
+        self.begin
+    }
+
+    pub fn end(&self) -> usize {
+        self.end
+    }
+
+    pub fn len(&self) -> usize {
+        self.end - self.begin
+    }
+
+    pub fn intersect(mut self, other: &RowSelector) -> Self {
+        self.values.and_inplace(&other.values);
+        RowSelector::new(
+            self.values,
+            max(self.begin, other.begin),
+            min(self.end, other.end),
+        )
+    }
+
+    pub fn concatenate(mut self, other: &RowSelector) -> Self {
+        assert_eq!(
+            self.end, other.begin,
+            "Can only concatenate consecutive selectors"
+        );
+        self.values.or_inplace(&other.values);
+        RowSelector::new(
+            self.values,
+            min(self.begin, other.begin),
+            max(self.end, other.end),
+        )
+    }
+
+    pub fn filter_array(&self, array: impl AsRef<Array>) -> VortexResult<Option<Array>> {
+        if self.begin != 0 {
+            vortex_bail!("Cannot filter with selections that don't start at offset 0")
+        }
+
+        let true_count = self.values.cardinality();
+        if true_count == 0 {
+            return Ok(None);
+        }
+
+        let array = array.as_ref();
+
+        if true_count == array.len() as u64 {
+            return Ok(Some(array.clone()));
+        }
+
+        let bitset = self
+            .values
+            .to_bitset()
+            .ok_or_else(|| vortex_err!("Couldn't create bitset for RowSelection"))?;
+
+        let byte_length = self.len().div_ceil(8);
+        let mut buffer = MutableBuffer::with_capacity(byte_length);
+        buffer.extend_from_slice(bitset.as_slice());
+        if byte_length > bitset.size_in_bytes() {
+            buffer.extend_zeros(byte_length - bitset.size_in_bytes());
+        }
+        let predicate = BoolArray::try_new(
+            BooleanBuffer::new(buffer.into(), 0, self.len()),
+            Validity::NonNullable,
+        )?;
+        filter(array, predicate).map(Some)
+    }
+
+    pub fn offset(self, offset: i64) -> RowSelector {
```

Comment:
with_offset

Reply:
Roaringbitmap calls this add_offset since we will actually shift the values. Maybe we should use the same name

Reply:
Or `shift` is in the python one. But I agree that Rust tends to prefix `with_`